tests/pgshovel/replication/targets/postgres.py

```diff
@@ -0,0 +1,309 @@
+            ],
+        ),
+        timestamp=Timestamp(seconds=0, nanos=0),
+        transaction=transaction_id,
+    )
+
+
+@pytest.yield_fixture
+def source_connection():
+    dsn = create_temporary_database('source')
+    yield psycopg2.connect(dsn)
+
+
+@pytest.yield_fixture
+def target_connection(target_cluster):
+    dsn = create_temporary_database('target')
+    _, connection = get_managed_databases(target_cluster, (dsn,)).items()[0]
+    yield connection
+
+
+@pytest.yield_fixture
+def target_cluster():
+    cluster = Cluster(
+        'test_%s' % (uuid.uuid1().hex,),
+        KazooClient('zookeeper'),
+    )
+
+    with cluster:
+        initialize_cluster(cluster)
+        yield cluster
+
+
+@pytest.yield_fixture
+def set(cluster, source_connection):
+    name = str(uuid.uuid1())
+    replication_set = ReplicationSetConfiguration()
+    replication_set.database.dsn = source_connection.dsn
+    replication_set.tables.add(name='auth_user', primary_keys=['id'])
+    create_set(cluster, name, replication_set)
+    yield name
+
+
+@pytest.yield_fixture
+def loader(cluster, set):
+    yield SimpleLoader(cluster, set)
+
+
+@pytest.yield_fixture
+def stream(cluster, set, client):
+    topic = str(uuid.uuid4())
+    client.ensure_topic_exists(topic)
+    yield LimitedKafkaStream(2, cluster, set, ('kafka:9092',), topic)
+
+
+@pytest.yield_fixture
+def target(target_cluster, set, target_connection):
+    yield PostgreSQLTarget(target_cluster, set, target_connection)
+
+
+@pytest.yield_fixture
+def client():
+    yield KafkaClient('kafka:9092')
+
+
+@pytest.yield_fixture
+def writer(client, stream):
+    producer = SimpleProducer(client)
+    yield KafkaWriter(producer, stream.topic)
+
+
+@pytest.yield_fixture
+def source_transaction_snapshot(source_connection):
+    with source_connection as conn, conn.cursor() as cursor:
+        cursor.execute('SELECT txid_current_snapshot();')
+        row = cursor.fetchone()
+        yield to_snapshot(row[0])
+
+
+@pytest.yield_fixture
+def invisible_transaction(source_transaction_snapshot):
+    # TODO: Figure out if we need to set the min, max here.
+    begin = BeginOperation(
+        start=Tick(
+            id=1,
+            snapshot=Snapshot(min=100, max=200),
+            timestamp=Timestamp(seconds=0, nanos=0),
+        ),
+        end=Tick(
+            id=2,
+            snapshot=Snapshot(min=150, max=250),
+            timestamp=Timestamp(seconds=10, nanos=0),
+        ),
+    )
+
+    mutation = get_mutation(source_transaction_snapshot.max + 1000)
+
+    commit = CommitOperation()
+
+    generator = make_batch_messages(
+        batch_identifier,
+        (
+            {'begin_operation': begin},
+            {'mutation_operation': mutation},
+            {'commit_operation': commit},
+        )
+    )
+
+    yield list(generator)
+
+
+@pytest.yield_fixture
+def follow_up_invisible_transaction(invisible_transaction):
+    operations = deepcopy(invisible_transaction)
```

Comment:
I feel bad about this `deepcopy`, but it's the easiest way to reuse existing message groups as a transaction, while having each message be its own object (so changes here don't effect the `invisible_transaction`).  Helpful generators to generate streams of individual instances of messages is one of the things I want to clean up.